Dependabot NuGet updater fails with InvalidCastException when using Central Package Management with transitive pinning

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package ecosystem

nuget

### Package manager version

_No response_

### Language version

dotnet 10

### Manifest location and content before the Dependabot update

**Summary:**
Dependabot fails while updating NuGet dependencies in a repository that uses Central Package Management (Directory.Packages.props) with CentralPackageTransitivePinningEnabled=true. The updater crashes with an InvalidCastException when attempting to write changes to a .csproj file.

**Environment:**
Package manager: NuGet
Using Central Package Management (ManagePackageVersionsCentrally=true)
CentralPackageTransitivePinningEnabled=true
Project files (.csproj) do not contain explicit package versions
Versions are managed exclusively in Directory.Packages.props

**Error:**
System.InvalidCastException: Unable to cast object of type 'Microsoft.Language.Xml.XmlDeclarationSyntax' to type 'Microsoft.Language.Xml.IXmlElementSyntax'
Occurs in:
NuGetUpdater.Core.Updater.FileWriters.XmlFileWriter.UpdatePackageVersionsAsync

**Additional Findings**:
1. When CentralPackageTransitivePinningEnabled is set to false:
- Dependabot succeeds.
- It creates a grouped PR.
- Updates are correctly applied to Directory.Packages.props & packages.lock.json

2. However, in some cases, Dependabot still adds VersionOverride entries in .csproj files:
   `<PackageReference Include="Azure.Messaging.ServiceBus" VersionOverride="7.20.1" />`
    even though the repository uses central version management.

**Conclusion:**
- The failure appears tied to handling of transitive pinning + VersionOverride logic.
- Dependabot incorrectly attempts to modify .csproj instead of relying solely on Directory.Packages.props.
- The XML writer crashes when processing the file structure during this operation.

**Impact:**
- Dependabot PRs fail entirely when transitive pinning is enabled.
- Requires disabling CentralPackageTransitivePinningEnabled or manually managing updates.

**Workaround:**
Set:
`<CentralPackageTransitivePinningEnabled>false</CentralPackageTransitivePinningEnabled>`
to allow Dependabot to proceed.

**Request:**
- Proper support for Central Package Management with CentralPackageTransitivePinningEnabled=true
- Avoid injecting VersionOverride into .csproj when versions are centrally managed
- Fix XML writer crash when handling such scenarios

### dependabot.yml content
```
version: 2

registries:
  nuget-feed:
    type: nuget-feed
    url: <private_artifact_url>
    username: "AzureDevOps"
    password: ${{ ALGO_FEED_PAT }}

updates:
  - package-ecosystem: "nuget"
    schedule:
      interval: "weekly"
    registries:
      - nuget-feed  
    directories:
      - /dir/
    labels:
      - "nuget"
      - "dependencies"
    open-pull-requests-limit: 10
    commit-message:
      prefix: "nuget"
      prefix-development: "nuget"
      include: "scope-and-version"
      separator: "-"
    # groups:
    #   nuget:
    #     patterns:
    #       - "*" 
```

### Updated dependency

- Dependabot successfully completes dependency discovery.
- It identifies a dependency update (e.g., Ardalis.SmartEnum.EFCore).
- It attempts to apply the update by adding a new element using VersionOverride.
- During XML write, the updater crashes with the above exception.

### What you expected to see, versus what you actually saw

- Dependabot should update the dependency version in Directory.Packages.props.
- It should not attempt to inject VersionOverride into .csproj files when versions are centrally managed.

### Native package manager behavior
_No response_

### Images of the diff or a link to the PR, issue, or logs
```
2026-03-18T11:40:25.6321134Z updater | 2026/03/18 11:40:25 ERROR Error type: unknown_error
2026-03-18T11:40:25.6321357Z - error-class: InvalidCastException
2026-03-18T11:40:25.6321622Z - error-message: System.InvalidCastException: Unable to cast object of type 'Microsoft.Language.Xml.XmlDeclarationSyntax' to type 'Microsoft.Language.Xml.IXmlElementSyntax'.
2026-03-18T11:40:25.6322130Z    at NuGetUpdater.Core.Updater.FileWriters.XmlFileWriter.UpdatePackageVersionsAsync(DirectoryInfo repoContentsPath, ImmutableArray`1 relativeFilePaths, ImmutableArray`1 originalDependencies, ImmutableArray`1 requiredPackageVersions, Boolean addPackageReferenceElementForPinnedPackages) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/XmlFileWriter.cs:line 270
2026-03-18T11:40:25.6322806Z    at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.TryPerformFileWritesAsync(IFileWriter fileWriter, DirectoryInfo repoContentsPath, DirectoryInfo originalDiscoveryDirectory, ProjectDiscoveryResult projectDiscovery, ImmutableArray`1 requiredPackageVersions) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 412
2026-03-18T11:40:25.6323432Z    at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.ProcessPackageReferenceUpdatesAsync(DirectoryInfo repoContentsPath, DirectoryInfo initialProjectDirectory, FileInfo projectPath, String dependencyName, NuGetVersion newDependencyVersion) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 258
2026-03-18T11:40:25.6324030Z    at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.RunAsync(DirectoryInfo repoContentsPath, FileInfo projectPath, String dependencyName, NuGetVersion oldDependencyVersion, NuGetVersion newDependencyVersion) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 49
2026-03-18T11:40:25.6324556Z    at NuGetUpdater.Core.UpdaterWorker.RunAsync(String repoRootPath, String workspacePath, String dependencyName, String previousDependencyVersion, String newDependencyVersion, Boolean isTransitive) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:line 84
2026-03-18T11:40:25.6325219Z    at NuGetUpdater.Core.Run.UpdateHandlers.GroupUpdateAllVersionsHandler.RunUngroupedDependencyUpdates(Job job, DirectoryInfo originalRepoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, IDiscoveryWorker discoveryWorker, IAnalyzeWorker analyzeWorker, IUpdaterWorker updaterWorker, IApiHandler apiHandler, ExperimentsManager experimentsManager, ILogger logger) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/UpdateHandlers/GroupUpdateAllVersionsHandler.cs:line 232
2026-03-18T11:40:25.6326272Z    at NuGetUpdater.Core.Run.UpdateHandlers.GroupUpdateAllVersionsHandler.HandleAsync(Job job, DirectoryInfo originalRepoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, IDiscoveryWorker discoveryWorker, IAnalyzeWorker analyzeWorker, IUpdaterWorker updaterWorker, IApiHandler apiHandler, ExperimentsManager experimentsManager, ILogger logger) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/UpdateHandlers/GroupUpdateAllVersionsHandler.cs:line 46
2026-03-18T11:40:25.6326948Z    at NuGetUpdater.Core.Run.RunWorker.RunScenarioHandlersWithErrorHandlingAsync(Job job, DirectoryInfo repoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, ExperimentsManager experimentsManager) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/RunWorker.cs:line 79
2026-03-18T11:40:25.6327282Z - package-manager: nuget
2026-03-18T11:40:25.6327427Z - job-id: 1042017906
2026-03-18T11:40:25.6504140Z [36mUpdate)job error: unknown_error {"error-class":"InvalidCastException","error-message":"System.InvalidCastException: Unable to cast object of type 'Microsoft.Language.Xml.XmlDeclarationSyntax' to type 'Microsoft.Language.Xml.IXmlElementSyntax'.\n   at NuGetUpdater.Core.Updater.FileWriters.XmlFileWriter.UpdatePackageVersionsAsync(DirectoryInfo repoContentsPath, ImmutableArray`1 relativeFilePaths, ImmutableArray`1 originalDependencies, ImmutableArray`1 requiredPackageVersions, Boolean addPackageReferenceElementForPinnedPackages) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/XmlFileWriter.cs:line 270\n   at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.TryPerformFileWritesAsync(IFileWriter fileWriter, DirectoryInfo repoContentsPath, DirectoryInfo originalDiscoveryDirectory, ProjectDiscoveryResult projectDiscovery, ImmutableArray`1 requiredPackageVersions) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 412\n   at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.ProcessPackageReferenceUpdatesAsync(DirectoryInfo repoContentsPath, DirectoryInfo initialProjectDirectory, FileInfo projectPath, String dependencyName, NuGetVersion newDependencyVersion) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 258\n   at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.RunAsync(DirectoryInfo repoContentsPath, FileInfo projectPath, String dependencyName, NuGetVersion oldDependencyVersion, NuGetVersion newDependencyVersion) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 49\n   at NuGetUpdater.Core.UpdaterWorker.RunAsync(String repoRootPath, String workspacePath, String dependencyName, String previousDependencyVersion, String newDependencyVersion, Boolean isTransitive) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:line 84\n   at NuGetUpdater.Core.Run.UpdateHandlers.GroupUpdateAllVersionsHandler.RunUngroupedDependencyUpdates(Job job, DirectoryInfo originalRepoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, IDiscoveryWorker discoveryWorker, IAnalyzeWorker analyzeWorker, IUpdaterWorker updaterWorker, IApiHandler apiHandler, ExperimentsManager experimentsManager, ILogger logger) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/UpdateHandlers/GroupUpdateAllVersionsHandler.cs:line 232\n   at NuGetUpdater.Core.Run.UpdateHandlers.GroupUpdateAllVersionsHandler.HandleAsync(Job job, DirectoryInfo originalRepoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, IDiscoveryWorker discoveryWorker, IAnalyzeWorker analyzeWorker, IUpdaterWorker updaterWorker, IApiHandler apiHandler, ExperimentsManager experimentsManager, ILogger logger) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/UpdateHandlers/GroupUpdateAllVersionsHandler.cs:line 46\n   at NuGetUpdater.Core.Run.RunWorker.RunScenarioHandlersWithErrorHandlingAsync(Job job, DirectoryInfo repoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, ExperimentsManager experimentsManager) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/RunWorker.cs:line 79","error-backtrace":"","package-manager":"nuget","job-id":"1042017906"}[39m
2026-03-18T11:40:25.6511541Z [36mUpdate unknown )job error: unknown_error {"error-class":"InvalidCastException","error-message":"System.InvalidCastException: Unable to cast object of type 'Microsoft.Language.Xml.XmlDeclarationSyntax' to type 'Microsoft.Language.Xml.IXmlElementSyntax'.\n   at NuGetUpdater.Core.Updater.FileWriters.XmlFileWriter.UpdatePackageVersionsAsync(DirectoryInfo repoContentsPath, ImmutableArray`1 relativeFilePaths, ImmutableArray`1 originalDependencies, ImmutableArray`1 requiredPackageVersions, Boolean addPackageReferenceElementForPinnedPackages) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/XmlFileWriter.cs:line 270\n   at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.TryPerformFileWritesAsync(IFileWriter fileWriter, DirectoryInfo repoContentsPath, DirectoryInfo originalDiscoveryDirectory, ProjectDiscoveryResult projectDiscovery, ImmutableArray`1 requiredPackageVersions) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 412\n   at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.ProcessPackageReferenceUpdatesAsync(DirectoryInfo repoContentsPath, DirectoryInfo initialProjectDirectory, FileInfo projectPath, String dependencyName, NuGetVersion newDependencyVersion) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 258\n   at NuGetUpdater.Core.Updater.FileWriters.FileWriterWorker.RunAsync(DirectoryInfo repoContentsPath, FileInfo projectPath, String dependencyName, NuGetVersion oldDependencyVersion, NuGetVersion newDependencyVersion) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/FileWriterWorker.cs:line 49\n   at NuGetUpdater.Core.UpdaterWorker.RunAsync(String repoRootPath, String workspacePath, String dependencyName, String previousDependencyVersion, String newDependencyVersion, Boolean isTransitive) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:line 84\n   at NuGetUpdater.Core.Run.UpdateHandlers.GroupUpdateAllVersionsHandler.RunUngroupedDependencyUpdates(Job job, DirectoryInfo originalRepoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, IDiscoveryWorker discoveryWorker, IAnalyzeWorker analyzeWorker, IUpdaterWorker updaterWorker, IApiHandler apiHandler, ExperimentsManager experimentsManager, ILogger logger) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/UpdateHandlers/GroupUpdateAllVersionsHandler.cs:line 232\n   at NuGetUpdater.Core.Run.UpdateHandlers.GroupUpdateAllVersionsHandler.HandleAsync(Job job, DirectoryInfo originalRepoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, IDiscoveryWorker discoveryWorker, IAnalyzeWorker analyzeWorker, IUpdaterWorker updaterWorker, IApiHandler apiHandler, ExperimentsManager experimentsManager, ILogger logger) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/UpdateHandlers/GroupUpdateAllVersionsHandler.cs:line 46\n   at NuGetUpdater.Core.Run.RunWorker.RunScenarioHandlersWithErrorHandlingAsync(Job job, DirectoryInfo repoContentsPath, DirectoryInfo caseInsensitiveRepoContentsPath, String baseCommitSha, ExperimentsManager experimentsManager) in /opt/nuget/lib/NuGetUpdater/NuGetUpdater.Core/Run/RunWorker.cs:line 79","error-backtrace":"","package-manager":"nuget","job-id":"1042017906"}[39m
2026-03-18T11:40:30.8386356Z ##[error]Error running updater: The updater encountered one or more errors.
2026-03-18T11:40:30.8532741Z ##[section]Finishing: Run Dependabot
```
### Smallest manifest that reproduces the issue

_No response_

--- a/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/FileWriters/XmlFileWriterTests.cs
+++ b/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/FileWriters/XmlFileWriterTests.cs
@@ -1949,4 +1949,52 @@ public class XmlFileWriterTests : FileWriterTestsBase
             ]
         );
     }
+
+    [Fact]
+    public async Task UpdatingAPinnedCentrallyManagedPackageUpdatesJustTheVersionNumberWhenDeclarationIsPresent()
+    {
+        await TestAsync(
+            useCentralPackageTransitivePinning: true,
+            files: [
+                ("src/project.csproj", """
+                    <?xml version="1.0"?>
+                    <Project Sdk="Microsoft.NET.Sdk">
+                      <ItemGroup>
+                        <PackageReference Include="Unrelated.Dependency" />
+                      </ItemGroup>
+                    </Project>
+                    """),
+                ("Directory.Packages.props", """
+                    <?xml version="1.0"?>
+                    <Project>
+                      <ItemGroup>
+                        <PackageVersion Include="Some.Dependency" Version="1.0.0" />
+                        <PackageVersion Include="Unrelated.Dependency" Version="3.0.0" />
+                      </ItemGroup>
+                    </Project>
+                    """)
+            ],
+            initialProjectDependencyStrings: ["Some.Dependency/1.0.0"],
+            requiredDependencyStrings: ["Some.Dependency/2.0.0"],
+            expectedFiles: [
+                ("src/project.csproj", """
+                    <?xml version="1.0"?>
+                    <Project Sdk="Microsoft.NET.Sdk">
+                      <ItemGroup>
+                        <PackageReference Include="Unrelated.Dependency" />
+                      </ItemGroup>
+                    </Project>
+                    """),
+                ("Directory.Packages.props", """
+                    <?xml version="1.0"?>
+                    <Project>
+                      <ItemGroup>
+                        <PackageVersion Include="Some.Dependency" Version="2.0.0" />
+                        <PackageVersion Include="Unrelated.Dependency" Version="3.0.0" />
+                      </ItemGroup>
+                    </Project>
+                    """)
+            ]
+        );
+    }
 }

--- a/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/XmlFileWriter.cs
+++ b/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/XmlFileWriter.cs
@@ -253,21 +253,47 @@ public class XmlFileWriter : IFileWriter
                     .ToArray();
                 if (matchingPackageVersionElementsAndPaths.Length > 0)
                 {
-                    // found matching `<PackageVersion>` element; if `Version` attribute is appropriate we're done, otherwise set `VersionOverride` attribute on new element
+                    // found matching `<PackageVersion>` element
                     var (matchingPackageVersionElement, filePath) = matchingPackageVersionElementsAndPaths.First();
                     var versionAttribute = matchingPackageVersionElement.GetAttributeCaseInsensitive(VersionMetadataName);
+                    var isVersionOverrideNeeded = false;
                     if (versionAttribute is not null &&
-                        VersionRange.TryParse(versionAttribute.Value, out var existingVersionRange) &&
-                        existingVersionRange.MinVersion == requiredVersion)
+                        VersionRange.TryParse(versionAttribute.Value, out var existingVersionRange))
                     {
-                        // version matches; no update needed
-                        _logger.Info($"Dependency {requiredPackageVersion.Name} already set to {requiredVersion}; no override needed.");
+                        if (existingVersionRange.MinVersion == requiredVersion)
+                        {
+                            // version matches; no update needed
+                            _logger.Info($"Dependency {requiredPackageVersion.Name} already set to {requiredVersion} in file {filePath}; no update needed.");
+                        }
+                        else if (existingVersionRange.Satisfies(oldVersion))
+                        {
+                            // found matching old version; update the attribute directly
+                            _logger.Info($"Dependency {requiredPackageVersion.Name} updated from version {oldVersion} to {requiredVersion} in file {filePath}.");
+                            ReplaceNode(
+                                filePath,
+                                matchingPackageVersionElement.AsNode,
+                                matchingPackageVersionElement.ReplaceAttribute(
+                                    versionAttribute,
+                                    versionAttribute.WithValue(requiredVersion.ToString())
+                                ).AsNode
+                            );
+                        }
+                        else
+                        {
+                            // version doesn't match; use `VersionOverride` attribute on new element
+                            isVersionOverrideNeeded = true;
+                        }
                     }
                     else
                     {
-                        // version doesn't match; use `VersionOverride` attribute on new element
-                        _logger.Info($"Dependency {requiredPackageVersion.Name} set to {requiredVersion}; using `{VersionOverrideMetadataName}` attribute on new element.");
-                        newElement = (IXmlElementSyntax)ReplaceNode(
+                        // version not found; use `VersionOverride` attribute on new element
+                        isVersionOverrideNeeded = true;
+                    }
+
+                    if (isVersionOverrideNeeded)
+                    {
+                        _logger.Info($"Dependency {requiredPackageVersion.Name} set to {requiredVersion} using `{VersionOverrideMetadataName}` attribute on new element in file {projectRelativePath}.");
+                        ReplaceNode(
                             projectRelativePath,
                             newElement.AsNode,
                             newElement.WithAttribute(VersionOverrideMetadataName, requiredVersion.ToString()).AsNode

